modules/imgproc/src/contours_link.cpp

```diff
@@ -0,0 +1,405 @@
+                }
+                else
+                {
+                    connect_flag = ICV_CONNECTING_BELOW;
+                    prev_point = rns[lower_run].next;
+                    n++;
+                    lower_run = rns[rns[lower_run].next].next;
+                }
+            }
+            break;
+        case ICV_CONNECTING_BELOW:
+            if( rns[lower_run].pt.x > rns[rns[upper_run].next].pt.x +1 )
+            {
+                rns[rns[upper_run].next].link = prev_point;
+                connect_flag = ICV_SINGLE;
+                k++;
+                upper_run = rns[rns[upper_run].next].next;
+            }
+            else
+            {
+                // First point of contour
+                int_rns.push_back(lower_run);
+
+                rns[lower_run].link = prev_point;
+                if( rns[rns[lower_run].next].pt.x < rns[rns[upper_run].next].pt.x )
+                {
+                    n++;
+                    prev_point = rns[lower_run].next;
+                    lower_run = rns[rns[lower_run].next].next;
+                }
+                else
+                {
+                    connect_flag = ICV_CONNECTING_ABOVE;
+                    k++;
+                    prev_point = rns[upper_run].next;
+                    upper_run = rns[rns[upper_run].next].next;
+                }
+            }
+            break;
+        }
+    }// k, n
+
+    for( ; n < lower_total/2; n++ )
+    {
+        if( connect_flag != ICV_SINGLE )
+        {
+            rns[prev_point].link = rns[lower_run].next;
+            connect_flag = ICV_SINGLE;
+            lower_run = rns[rns[lower_run].next].next;
+            continue;
+        }
+        rns[rns[lower_run].next] = rns[rns[lower_run].next];
+        rns[lower_run].link = rns[lower_run].next;
+
+        //First point of contour
+        ext_rns.push_back(lower_run);
+        lower_run = rns[rns[lower_run].next].next;
+    }
+
+    for( ; k < upper_total/2; k++ )
+    {
+        if( connect_flag != ICV_SINGLE )
+        {
+            rns[rns[upper_run].next].link = prev_point;
+            connect_flag = ICV_SINGLE;
+            upper_run = rns[rns[upper_run].next].next;
+            continue;
+        }
+        rns[rns[upper_run].next] = rns[rns[upper_run].next];
+        rns[rns[upper_run].next].link = upper_run;
+        upper_run = rns[rns[upper_run].next].next;
+    }
+}
+
+
+void LinkRunner::process(Mat & image)
+{
+    const Size sz = image.size();
+    int j;
+    int  lower_total;
+    int  upper_total;
+    int  all_total;
+
+    Point cur_point;
+
+    // First line. None of runs is binded
+    rns.push_back(LRP());
+    int upper_line = (int)rns.size() - 1;
+    int cur = upper_line;
+    for( j = 0; j < sz.width; )
+    {
+        j = findStartContourPoint(image.ptr<uchar>(), sz, j);
+
+        if( j == sz.width )
+            break;
+
+        cur_point.x = j;
+
+        rns.push_back(LRP(cur_point));
```

Comment:
Looks like rns has predictable size. It makes sense to use reserve and emplace_back.

Reply:
`rns` size is not predictable, it depends on number of contours on image and their shape. We can reserve some amount assuming there are at least some contours though.